<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>CHR:POS to RSID mapping • genepi.utils</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CHR:POS to RSID mapping">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genepi.utils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.31</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a></li>
    <li><a class="dropdown-item" href="../articles/clumping.html">Clumping</a></li>
    <li><a class="dropdown-item" href="../articles/collider_bias.html">Collider bias</a></li>
    <li><a class="dropdown-item" href="../articles/ld_matrix.html">LD matrix</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Plotting</a></li>
    <li><a class="dropdown-item" href="../articles/rationale.html">Rationale and key points</a></li>
    <li><a class="dropdown-item" href="../articles/standardise_gwas.html">Standardise GWAS</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>CHR:POS to RSID mapping</h1>
            
      

      <div class="d-none name"><code>chrpos_to_rsid.Rmd</code></div>
    </div>

    
    
<p>Here is an example of mapping CHR:POS:REF:ALT columns to dbSNP build
156 RSIDs.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nicksunderland.github.io/genepi.utils/">genepi.utils</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt;    chromosome base_pair_location effect_allele other_allele       RSID</span></span>
<span><span class="co">#&gt;         &lt;int&gt;              &lt;int&gt;        &lt;char&gt;       &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:         10          100000625             G            A  rs7899632</span></span>
<span><span class="co">#&gt; 2:         10          100000645             C            A rs61875309</span></span>
<span><span class="co">#&gt; 3:         10          100003242             G            T rs12258651</span></span>
<span><span class="co">#&gt; 4:         10          100003785             C            T  rs1359508</span></span>
<span><span class="co">#&gt; 5:         10          100004360             A            G  rs1048754</span></span></code></pre></div>
<div class="section level2">
<h2 id="available-builds">Available builds<a class="anchor" aria-label="anchor" href="#available-builds"></a>
</h2>
<p>Currently we have the latest dbSNP build 156 available for hg37 and
hg38.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/which_dbsnp_builds.html">which_dbsnp_builds</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="monitoring-progress">Monitoring progress<a class="anchor" aria-label="anchor" href="#monitoring-progress"></a>
</h2>
<p>The mapping can take some time, depending on the number of cores on
your machine. To monitor progress we have to use the form below. We need
to wrap the function in <code>progressr::with_progress({  })</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                  chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                  pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                  ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                  nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                  build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Chromosomes to process:  10</span></span>
<span><span class="co"># |=================                                                                    |  12%</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="function-options">Function options<a class="anchor" aria-label="anchor" href="#function-options"></a>
</h2>
<div class="section level3">
<h3 id="alt-rsid-output">Alt RSID output<a class="anchor" aria-label="anchor" href="#alt-rsid-output"></a>
</h3>
<p>In the dbSNP database there are multiple RSID assigned to the same
CHR:POS:REF:ALT combination. This function will return the first RSID
encountered and filter out the rest. If you want to get the alternative
RSIDs then use the <code>alt_rsids=TRUE</code> flag. This will change
the structure of the output to a named <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> of two
<code>data.table</code> elements: 1) “data”=gwas_data and 2)
“alt_rsids”=alt_rsid_data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span>, </span>
<span>                                                alt_rsids<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="allele-specification-matching">Allele specification / matching<a class="anchor" aria-label="anchor" href="#allele-specification-matching"></a>
</h3>
<p>Although matching on alleles is desirable, especially for indels, we
can still match on just chromosome and position by either leaving
<code>ea_col</code> and <code>nea_col</code> arguments out, or setting
them to <code>NULL</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span></code></pre></div>
<p>Alleles are not always aligned to the reference and so we allow
allele flipping by default with the parameter <code>flip="allow"</code>.
If you would like to flag which variants have been assigned an rsID
based on flipped alleles use <code>flip="report"</code>; this will add a
logical column to the output called <code>rsid_flip_match</code>. To
prevent allele flipping and only assign based on the reference allele
orientation use <code>flip="no_flip"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                flip    <span class="op">=</span> <span class="st">"report"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span></code></pre></div>
<p>Coding of indels as D/I is allowed, however please check the
alternate rsID output as mutliple rsIDs (indels) may be present for D/I
coding at any particular base position.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_id_coding</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example3_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span>dt      <span class="op">=</span> <span class="va">gwas_id_coding</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                flip    <span class="op">=</span> <span class="st">"allow"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span>, </span>
<span>                                                alt_rsids<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="evaluation-speed">Evaluation speed<a class="anchor" aria-label="anchor" href="#evaluation-speed"></a>
</h2>
<p>The choice of parameters will have a small impact computation speed.
Since the dbSNP data is stored as <code>.fst</code> binary files and
only the desired rows / columns are ever read into memory, the more data
that is requested the slower the computation. That said, it is still
much faster / feasible than trying to read in the entire dbSNP database
(over 1 billion rsIDs).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshuaulrich/microbenchmark/" class="external-link">microbenchmark</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># some GWAS data</span></span>
<span><span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span> <span class="va">gwas_sumstats_4.3million_rows</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># benchmarking</span></span>
<span><span class="va">mbm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html" class="external-link">microbenchmark</a></span><span class="op">(</span><span class="st">"chrpos_to_rsid: no alleles, no alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                              <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                                <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                             chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                             pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                             ea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                             nea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                             build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                             alt_rsids<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                                             flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                              <span class="op">}</span><span class="op">)</span></span>
<span>                        <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: no alleles, with alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: with alleles, no alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="st">"EFFECT_ALLELE"</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="st">"OTHER_ALLELE"</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: with alleles, with alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="st">"EFFECT_ALLELE"</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="st">"OTHER_ALLELE"</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      times <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">mbm</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
