<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CHR:POS to RSID mapping • genepi.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CHR:POS to RSID mapping">
<meta property="og:description" content="genepi.utils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CHR:POS to RSID mapping</h1>
            
      
      
      <div class="hidden name"><code>chrpos_to_rsid.Rmd</code></div>

    </div>

    
    
<p>Here is an example of mapping CHR:POS:REF:ALT columns to dbSNP build
156 RSIDs.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genepi.utils</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span>
<span><span class="co">#&gt;    chromosome base_pair_location effect_allele other_allele</span></span>
<span><span class="co">#&gt; 1:         10          100000625             G            A</span></span>
<span><span class="co">#&gt; 2:         10          100000645             C            A</span></span>
<span><span class="co">#&gt; 3:         10          100002841            CT            C</span></span>
<span><span class="co">#&gt; 4:         10          100003785             C            T</span></span>
<span><span class="co">#&gt; 5:         10          100004360             A            G</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt;    chromosome base_pair_location effect_allele other_allele       RSID</span></span>
<span><span class="co">#&gt;         &lt;int&gt;              &lt;int&gt;        &lt;char&gt;       &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:         10          100000625             G            A  rs7899632</span></span>
<span><span class="co">#&gt; 2:         10          100000645             C            A rs61875309</span></span>
<span><span class="co">#&gt; 3:         10          100003242             G            T rs12258651</span></span>
<span><span class="co">#&gt; 4:         10          100003785             C            T  rs1359508</span></span>
<span><span class="co">#&gt; 5:         10          100004360             A            G  rs1048754</span></span></code></pre></div>
<div class="section level2">
<h2 id="available-builds">Available builds<a class="anchor" aria-label="anchor" href="#available-builds"></a>
</h2>
<p>Currently we have the latest dbSNP build 156 available for hg37 and
hg38.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/which_dbsnp_builds.html">which_dbsnp_builds</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                              b37_dbsnp156 </span></span>
<span><span class="co">#&gt; "/Users/xx20081/Documents/local_data/genome_reference/dbsnp/b37_dbsnp156" </span></span>
<span><span class="co">#&gt;                                                              b38_dbsnp156 </span></span>
<span><span class="co">#&gt; "/Users/xx20081/Documents/local_data/genome_reference/dbsnp/b38_dbsnp156"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="monitoring-progress">Monitoring progress<a class="anchor" aria-label="anchor" href="#monitoring-progress"></a>
</h2>
<p>The mapping can take some time, depending on the number of cores on
your machine. To monitor progress we have to use the form below. First
we set up the parallel processing plan with the number of workers - this
should be &lt;= the number of cores on your machine, which can be
queried with <code><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">parallel::detectCores()</a></code>. Then to monitor
progress in the console we need to wrap the function in
<code>progressr::with_progress({  })</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                  chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                  pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                  ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                  nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                  build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Chromosomes to process:  10</span></span>
<span><span class="co"># |=================                                                                    |  12%</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="function-options">Function options<a class="anchor" aria-label="anchor" href="#function-options"></a>
</h2>
<div class="section level3">
<h3 id="alt-rsid-output">Alt RSID output<a class="anchor" aria-label="anchor" href="#alt-rsid-output"></a>
</h3>
<p>In the dbSNP database there are multiple RSID assigned to the same
CHR:POS:REF:ALT combination. This function will return the first RSID
encountered and filter out the rest. If you want to get the alternative
RSIDs then use the <code>alt_rsids=TRUE</code> flag. This will change
the structure of the output to a named <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> of two
<code>data.table</code> elements: 1) “data”=gwas_data and 2)
“alt_rsids”=alt_rsid_data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span>, </span>
<span>                                                alt_rsids<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (5/5), of which 0% (0/5) where found flipping alleles</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ data     :Classes 'data.table' and 'data.frame':  5 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;   ..$ RSID              : chr [1:5] "rs7899632" "rs61875309" "rs146107628" "rs1359508" ...</span></span>
<span><span class="co">#&gt;   ..$ chromosome        : int [1:5] 10 10 10 10 10</span></span>
<span><span class="co">#&gt;   ..$ base_pair_location: int [1:5] 100000625 100000645 100002841 100003785 100004360</span></span>
<span><span class="co">#&gt;   ..$ effect_allele     : chr [1:5] "G" "C" "CT" "C" ...</span></span>
<span><span class="co">#&gt;   ..$ other_allele      : chr [1:5] "A" "A" "C" "T" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;   ..- attr(*, "orig_type")= chr "data.table"</span></span>
<span><span class="co">#&gt;  $ alt_rsids:Classes 'data.table' and 'data.frame':  0 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ RSID    : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ CHR     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ BP      : int(0) </span></span>
<span><span class="co">#&gt;   ..$ REF     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ ALT     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ baseRSID: chr(0) </span></span>
<span><span class="co">#&gt;   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;   ..- attr(*, "orig_type")= chr "data.table"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="allele-specification-matching">Allele specification / matching<a class="anchor" aria-label="anchor" href="#allele-specification-matching"></a>
</h3>
<p>Although matching on alleles is desirable, especially for indels, we
can still match on just chromosome and position by either leaving
<code>ea_col</code> and <code>nea_col</code> arguments out, or setting
them to <code>NULL</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (5/5), of which 0% (0/5) where found flipping alleles</span></span>
<span><span class="co">#&gt; Warning in `[.data.table`(out_data, , `:=`("rsid_flip_match", NULL)): Column</span></span>
<span><span class="co">#&gt; 'rsid_flip_match' does not exist to remove</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span><span class="co">#&gt; Classes 'data.table' and 'data.frame':   5 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ RSID              : chr  "rs7899632" "rs61875309" "rs146107628" "rs1359508" ...</span></span>
<span><span class="co">#&gt;  $ chromosome        : int  10 10 10 10 10</span></span>
<span><span class="co">#&gt;  $ base_pair_location: int  100000625 100000645 100002841 100003785 100004360</span></span>
<span><span class="co">#&gt;  $ effect_allele     : chr  "G" "C" "CT" "C" ...</span></span>
<span><span class="co">#&gt;  $ other_allele      : chr  "A" "A" "C" "T" ...</span></span>
<span><span class="co">#&gt;  - attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;  - attr(*, "orig_type")= chr "data.table"</span></span></code></pre></div>
<p>Alleles are not always aligned to the reference and so we allow
allele flipping by default with the parameter <code>flip="allow"</code>.
If you would like to flag which variants have been assigned an rsID
based on flipped alleles use <code>flip="report"</code>; this will add a
logical column to the output called <code>rsid_flip_match</code>. To
prevent allele flipping and only assign based on the reference allele
orientation use <code>flip="no_flip"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                flip    <span class="op">=</span> <span class="st">"report"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (5/5), of which 0% (0/5) where found flipping alleles</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span><span class="co">#&gt; Classes 'data.table' and 'data.frame':   5 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;  $ RSID              : chr  "rs7899632" "rs61875309" "rs146107628" "rs1359508" ...</span></span>
<span><span class="co">#&gt;  $ chromosome        : int  10 10 10 10 10</span></span>
<span><span class="co">#&gt;  $ base_pair_location: int  100000625 100000645 100002841 100003785 100004360</span></span>
<span><span class="co">#&gt;  $ effect_allele     : chr  "G" "C" "CT" "C" ...</span></span>
<span><span class="co">#&gt;  $ other_allele      : chr  "A" "A" "C" "T" ...</span></span>
<span><span class="co">#&gt;  $ rsid_flip_match   : logi  FALSE FALSE FALSE FALSE FALSE</span></span>
<span><span class="co">#&gt;  - attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;  - attr(*, "orig_type")= chr "data.table"</span></span></code></pre></div>
<p>Coding of indels as D/I is allowed, however please check the
alternate rsID output as mutliple rsIDs (indels) may be present for D/I
coding at any particular base position.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_id_coding</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example3_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gwas_with_rsids</span> <span class="op">&lt;-</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span>dt      <span class="op">=</span> <span class="va">gwas_id_coding</span>,</span>
<span>                                                chr_col <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>                                                pos_col <span class="op">=</span> <span class="st">"base_pair_location"</span>,</span>
<span>                                                ea_col  <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>                                                nea_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>                                                flip    <span class="op">=</span> <span class="st">"allow"</span>,</span>
<span>                                                build   <span class="op">=</span> <span class="st">"b37_dbsnp156"</span>, </span>
<span>                                                alt_rsids<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (5/5), of which 20% (1/5) where found flipping alleles</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gwas_with_rsids</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ data     :Classes 'data.table' and 'data.frame':  5 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;   ..$ RSID              : chr [1:5] "rs7899632" "rs61875309" "rs2135792004" "rs1359508" ...</span></span>
<span><span class="co">#&gt;   ..$ chromosome        : int [1:5] 10 10 10 10 10</span></span>
<span><span class="co">#&gt;   ..$ base_pair_location: int [1:5] 100000625 100000645 100002841 100003785 100004360</span></span>
<span><span class="co">#&gt;   ..$ effect_allele     : chr [1:5] "G" "C" "I" "C" ...</span></span>
<span><span class="co">#&gt;   ..$ other_allele      : chr [1:5] "A" "A" "D" "T" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;   ..- attr(*, "orig_type")= chr "data.table"</span></span>
<span><span class="co">#&gt;  $ alt_rsids:Classes 'data.table' and 'data.frame':  0 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ RSID    : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ CHR     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ BP      : int(0) </span></span>
<span><span class="co">#&gt;   ..$ REF     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ ALT     : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ baseRSID: chr(0) </span></span>
<span><span class="co">#&gt;   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;   ..- attr(*, "orig_type")= chr "data.table"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="evaluation-speed">Evaluation speed<a class="anchor" aria-label="anchor" href="#evaluation-speed"></a>
</h2>
<p>The choice of parameters will have a small impact computation speed.
Since the dbSNP data is stored as <code>.fst</code> binary files and
only the desired rows / columns are ever read into memory, the more data
that is requested the slower the computation. That said, it is still
much faster / feasible than trying to read in the entire dbSNP database
(over 1 billion rsIDs).</p>
<div class="figure" style="text-align: center">
<img src="figures/microbenchmark_chrpos_to_rsid.png" alt="Computation speed: Apple M2 Max 96GB 10 cores; 4.3 million SNPs queried against ~1 billion dbSNP156 RSIDs" width="100%"><p class="caption">
Computation speed: Apple M2 Max 96GB 10 cores; 4.3 million SNPs queried
against ~1 billion dbSNP156 RSIDs
</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshuaulrich/microbenchmark/" class="external-link">microbenchmark</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># some GWAS data</span></span>
<span><span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span> <span class="va">gwas_sumstats_4.3million_rows</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># benchmarking</span></span>
<span><span class="va">mbm</span> <span class="op">&lt;-</span> <span class="fu">microbenchmark</span><span class="op">(</span><span class="st">"chrpos_to_rsid: no alleles, no alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                              <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>                              <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                                <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                             chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                             pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                             ea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                             nea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                             build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                             alt_rsids<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                                             flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                              <span class="op">}</span><span class="op">)</span></span>
<span>                        <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: no alleles, with alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: with alleles, no alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="st">"EFFECT_ALLELE"</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="st">"OTHER_ALLELE"</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      <span class="st">"chrpos_to_rsid: with alleles, with alt rsids"</span> <span class="op">=</span> <span class="op">{</span></span>
<span>                        <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>                        <span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/with_progress.html" class="external-link">with_progress</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                          <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid</a></span><span class="op">(</span><span class="va">dt</span>,</span>
<span>                                                       chr_col<span class="op">=</span><span class="st">"CHR"</span>,</span>
<span>                                                       pos_col<span class="op">=</span><span class="st">"POS"</span>,</span>
<span>                                                       ea_col<span class="op">=</span><span class="st">"EFFECT_ALLELE"</span>,</span>
<span>                                                       nea_col<span class="op">=</span><span class="st">"OTHER_ALLELE"</span>,</span>
<span>                                                       build<span class="op">=</span><span class="st">"b37_dbsnp156"</span>,</span>
<span>                                                       alt_rsids<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                       flip<span class="op">=</span><span class="st">"report"</span><span class="op">)</span></span>
<span>                        <span class="op">}</span><span class="op">)</span></span>
<span>                      <span class="op">}</span>,</span>
<span>                      times <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">mbm</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
