<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clumping â€¢ genepi.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clumping">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.31</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/rationale.html">Rationale and key points</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Clumping</h1>
            
      

      <div class="hidden name"><code>clumping.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Clumping GWAS results aims to retain only independent variants,
accounting for linkage disequilibrium (highly correlated signals). The
<code><a href="../reference/clump.html">clump()</a></code> function uses PLINK2 in the background and requires
a genome reference dataset with linkage disequilibrium information. This
reference must contain the same variant IDs as your GWAS results. It is
recommended to first annotate your GWAS with rsIDs - this can be
achieved with the <code><a href="../reference/chrpos_to_rsid.html">chrpos_to_rsid()</a></code> function if your data is
not already annotated. The data is also expected in genepi.utils
standard form, so run <code>standardise_gwas()</code> if needed. The
standardisation and rsID annotation can also be performed in one step
with the <code>standardise_gwas()</code> function, by passing an
appropriate <code>populate_rsid</code> argument.</p>
<p>A common reference dataset is the 1000 genomes phase 3 which can be
downloaded from the PLINK (<a href="https://www.cog-genomics.org/plink/2.0/resources#phase3_1kg" class="external-link uri">https://www.cog-genomics.org/plink/2.0/resources#phase3_1kg</a>).
<strong>Important</strong>: if you use a different dataset ensure that
the IDs match those in your dataset).</p>
<p>To ready the PLINK2 pfiles for use with clumping you can run the
following, as there appear to be some duplicate rsIDs in this dataset
that PLINK2 does not like.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">plink2</span> <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  <span class="at">--pfile</span> all_phase3 <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="at">--rm-dup</span> force-first <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="at">--make-pgen</span> <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="at">--out</span> all_phase3_nodup</span></code></pre></div>
<p>The path to your processed reference files
(<code>path/to/plink_ref/all_phase3_nodup</code>) can be directly passed
the the <code><a href="../reference/clump.html">clump()</a></code> function as the <code>plink_ref</code>
argument, or, you can configure the path in the package by using
<code><a href="../reference/set_1000G_reference.html">set_1000G_reference()</a></code> - this only needs to be done once per
install. The currently set reference path can be queried with
<code><a href="../reference/which_1000G_reference.html">which_1000G_reference()</a></code>.</p>
<p>You will need PLINK2 installed on your computer. The
<code><a href="../reference/clump.html">clump()</a></code> function takes a path to the plink executable as
the <code>plink2</code> argument. Alternatively, you can configure the
path in the package by using <code><a href="../reference/set_plink2.html">set_plink2()</a></code> - this only needs
to be done once per install. The currently set PLINK2 path can be
queried with <code><a href="../reference/which_plink2.html">which_plink2()</a></code>. If the <code>plink2</code>
argument is set to <code>NULL</code> the function will attempt to call
plink directly, equivalent of running <code>plink2 --options</code>
directly in your command line.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nicksunderland.github.io/genepi.utils/">genepi.utils</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the gwas data</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>dat<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example2_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span>, map<span class="op">=</span><span class="st">"ns_map"</span>, fill_rsid<span class="op">=</span><span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.data.table.html">as.data.table</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># check path to reference</span></span>
<span><span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check path to plink2 executable</span></span>
<span><span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clumping">Clumping<a class="anchor" aria-label="anchor" href="#clumping"></a>
</h2>
<p>Running the clumping procedure will annotate your GWAS with two
columns:</p>
<ul>
<li>
<code>index</code>: a logical column indicating whether the variant
is an index/lead variant. If you only want clumped/index SNPs you should
filter on this e.g.Â <code>gwas[index==TRUE, ]</code>
</li>
<li>
<code>clump</code>: an integer column indicating the clump that the
variant belongs to.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># input gwas; n.b. trailing `[]` just for printing in rmarkdown</span></span>
<span><span class="va">gwas</span><span class="op">[</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># run clumping</span></span>
<span><span class="va">clumped_gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clump.html">clump</a></span><span class="op">(</span><span class="va">gwas</span>, </span>
<span>                      p1 <span class="op">=</span> <span class="fl">1.0</span>,</span>
<span>                      p2 <span class="op">=</span> <span class="fl">1.0</span>, </span>
<span>                      r2 <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>                      kb <span class="op">=</span> <span class="fl">250</span>, </span>
<span>                      plink2    <span class="op">=</span> <span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>                      plink_ref <span class="op">=</span> <span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span>build<span class="op">=</span><span class="st">"GRCh37"</span><span class="op">)</span>, </span>
<span>                      parallel_cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># view result</span></span>
<span><span class="va">clumped_gwas</span><span class="op">[</span><span class="op">]</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
