<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Collider bias • genepi.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Collider bias">
<meta property="og:description" content="genepi.utils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Collider bias</h1>
            
      
      
      <div class="hidden name"><code>collider_bias.Rmd</code></div>

    </div>

    
    
<p>An example of using the collider bias functions to investigate the
degree of index event (collider) bias in your GWAS progression data. We
start with independent variants from the incident GWAS of interest -
this will will require pre-processing with appropriate clumping
parameters. All variants from the progression GWAS are included in the
initial step.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genepi.utils</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read in the data</span></span>
<span><span class="va">gwas_clumped_incidence</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">clumped_incidence_path</span><span class="op">)</span><span class="op">[</span><span class="va">index</span><span class="op">==</span><span class="cn">TRUE</span>, <span class="op">]</span> <span class="co"># take only the index SNPs </span></span>
<span><span class="va">gwas_progression</span>       <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">progression_path</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="slope-hunter">Slope-hunter<a class="anchor" aria-label="anchor" href="#slope-hunter"></a>
</h2>
<div class="section level3">
<h3 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p>Run the SlopeHunter expectation-maximization method to estimate the
bias adjustment factor. The algorithm uses model-based clustering and
proposes that the distributions of the incidence(GI) and progression
(GP) BETAs can be written like so:</p>
<p><span class="math display">\[
  \binom{\beta_{GI}}{\beta'_{GP}}
  \sim
  \color{#F8766D}{
  \pi_{1}N
  \begin{pmatrix}
  \underline{0},
  \begin{bmatrix}
  \sigma_{I}^{2} &amp; b_{1}\sigma_{I}^{2} \\
  b_1\sigma_{I}^{2} &amp; b_{1}^{2}\sigma_{I}^{2} \\
  \end{bmatrix}
  \end{pmatrix}}
  +
  \color{#619CFF}{
  \pi_{2}N
  \begin{pmatrix}
  \underline{0},
  \begin{bmatrix}
  \sigma_{I}^{2} &amp; b_{1}\sigma_{I}^{2} + \sigma_{IP} \\
  b_1\sigma_{I}^{2} + \sigma_{IP} &amp; b_{1}^{2}\sigma_{I}^{2} +
\sigma_{P}^{2} + 2b_{1}\sigma_{IP} \\
  \end{bmatrix}
  \end{pmatrix}}
  \color{gray}{
  +
  \pi_{3}
  \begin{pmatrix}
  \eta_{0} \\
  N(0, \sigma_{P}^{2})
  \end{pmatrix}
  +
  \pi_{4}
  \begin{pmatrix}
  \eta_{0} \\
  \eta_{0}
  \end{pmatrix}}
\]</span></p>
<ul>
<li>Cluster 1: SNPs that cause incidence but not progression (red)<br>
</li>
<li>Cluster 2: SNPs that cause incidence and progression (blue)<br>
</li>
<li>Cluster 3: SNPs that cause progression but not incidence
(gray)<br>
</li>
<li>Cluster 4: SNPs that cause neither incidence or progression
(gray)</li>
</ul>
<p>The values: <span class="math display">\[{\pi_{1}, \pi_{2}, \pi_{3},
\pi_{4}}\]</span> … are the probabilities that a SNP belongs to the
respective clusters.</p>
<p>The first thing to note is that we can filter out SNPs from cluster 3
and 4 by only including SNPs with a significant association (P-value)
with incidence - this is the <code>ip</code> parameter. The problem is
then reduced to finding two clusters that best fit distributions 1 and 2
of the equation above. The SlopeHunter EM algorithm iteratively
determines which SNPs belong to each distribution and once complete the
adjustment factor (slope gradient) can be determined from group 1,
i.e. only those SNPs thought to solely cause incidence.</p>
<p>This function’s code is adapted from the SlopeHunter R package and if
using this method the SlopeHunter package should be referenced as the
original source.</p>
</div>
<div class="section level3">
<h3 id="run">Run<a class="anchor" aria-label="anchor" href="#run"></a>
</h3>
<p>Running the Slope-hunter algorithm requires initialising
parameters.</p>
<ul>
<li>
<code>ip</code> is the initial p-value by which to filter the
incident GWA variants. Variants with a p-value greater that
<code>ip</code> will be removed from the analysis (i.e. variants in
clusters 3 &amp; 4).</li>
<li>
<code>pi0</code> is an initial guess for the proportion of variants
that only associate with incidence (i.e. belong to cluster 1).<br>
</li>
<li>
<code>sxy1</code> is an initial guess for the covariance between
incidence and progression betas in cluster 2.</li>
</ul>
<p>The standard error of the slope is estimated with bootstrapping, 100
samples in the case below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slopehunter.html">slopehunter</a></span><span class="op">(</span>gwas_i     <span class="op">=</span> <span class="va">gwas_clumped_incidence</span>,</span>
<span>                      gwas_p     <span class="op">=</span> <span class="va">gwas_progression</span>,</span>
<span>                      merge      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span><span class="op">=</span><span class="st">"CHR"</span>, <span class="st">"BP"</span><span class="op">=</span><span class="st">"BP"</span><span class="op">)</span>, </span>
<span>                      ip         <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>                      pi0        <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                      sxy1       <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span>                      bootstraps <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualise">Visualise<a class="anchor" aria-label="anchor" href="#visualise"></a>
</h3>
<p>We can visualise the iterative EM algorithm like so. The variants are
allocated to either cluster 1 (incidence only) or cluster 2 (incidence
and progression) based on how well they conform to defined bivariate
distributions for each cluster. The key point is that the incidence-only
distribution has a linear relationship between incidence and progression
betas, with no constant term - i.e. we search for a straight line
passing through the origin. At each iteration the distribution
parameters are recalculated (variance / s.dev / cov) and variant
reassigned to either cluster 1 or 2, until the model converges.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_slopehunter_iters.html">plot_slopehunter_iters</a></span><span class="op">(</span>gwas_i <span class="op">=</span> <span class="va">gwas_clumped_incidence</span>,</span>
<span>                            gwas_p <span class="op">=</span> <span class="va">gwas_progression</span>,</span>
<span>                            ip     <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>                            pi0    <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                            sxy1   <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span>                            x_lims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.2</span>, <span class="fl">0.2</span><span class="op">)</span>, </span>
<span>                            y_lims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/slopehunter.gif" width="90%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="the-b-slope">The b-slope<a class="anchor" aria-label="anchor" href="#the-b-slope"></a>
</h2>
<p>In the expectation (E) part of the Slope-hunter algorithm we
initially set the covariance matrix of the first component: <span class="math display">\[
\color{#F8766D}{
  \pi_{1}N
  \begin{pmatrix}
  \underline{0},
  \begin{bmatrix}
  \sigma_{I}^{2} &amp; b_{1}\sigma_{I}^{2} \\
  b_1\sigma_{I}^{2} &amp; b_{1}^{2}\sigma_{I}^{2} \\
  \end{bmatrix}
  \end{pmatrix}}
\]</span> The off-diagonal components are set to
<code>sxy0 = sx0*sy0*dir0*0.95</code> where <code>sx0</code> and
<code>sy0</code> are the standard deviations of the incidence and
progression betas, respectively, and the <code>dir0</code> term is the
<code><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign()</a></code> of the covariance - which is positive in this
case:</p>
<div class="figure" style="text-align: center">
<img src="figures/slopehunter_cov_x_y.png" alt="Cov(incidence, progression)" width="90%"><p class="caption">
Cov(incidence, progression)
</p>
</div>
<p>In the maximisation (M) step of the algorithm, we update
<code>dir0</code> as the <code><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign()</a></code> of
<code>sum(pt*xbeta*ybeta)/sum(pt)</code>. <code>pt</code> is a positive
array and so the sign depends on the relative proportions of negatives
and positive of the product of <code>xbeta</code> and
<code>ybeta</code>. This is positive in our case, as visualised in this
histogram:</p>
<div class="figure" style="text-align: center">
<img src="figures/slopehunter_product_x_y.png" alt="product(incidence, progression)" width="90%"><p class="caption">
product(incidence, progression)
</p>
</div>
<p>The effect, I believe, is that a positive slope is maintained at each
iteration of the algorithm.</p>
</div>
<div class="section level2">
<h2 id="overall-assessment---multiple-methods">Overall assessment - multiple methods<a class="anchor" aria-label="anchor" href="#overall-assessment---multiple-methods"></a>
</h2>
<p>The algorithms can be sensitive to the incidence p-value threshold
<code>ip</code>. Too high and null variants without any effect on
incidence will be included. Too low and variants will small, but
potentially significant, effects on incidence will be removed. It is a
good idea to investigate the stability of the adjustment factor estimate
to difference in <code>ip</code> threshold.</p>
<p>We can run many combinations of the above analyses by defining
sets/vectors of parameters. Note: be careful with large numbers of
parameters as this will quickly lead to large numbers of combinations.
e.g. </p>
<ul>
<li>methods = c(“slopehunter”)<br>
</li>
<li>ip = c(0.05,0.001,0.00001)<br>
</li>
<li>pi0 = c(0.6, 0.65, 0.7).</li>
<li>sxy1 = c(1e-5, 1e-4, 1e-3)</li>
</ul>
<p>…will lead to 27 separate analyses.</p>
<p>We can use the <code><a href="../reference/analyse_collider_bias.html">analyse_collider_bias()</a></code> function to run
all combinations of methods and parameters. Here is the stability of the
b-slope with different p-values (<code>ip</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyse_collider_bias.html">analyse_collider_bias</a></span><span class="op">(</span>gwas_i <span class="op">=</span> <span class="va">gwas_clumped_incidence</span>,</span>
<span>                                 gwas_p <span class="op">=</span> <span class="va">gwas_progression</span>,</span>
<span>                                 merge  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span><span class="op">=</span><span class="st">"CHR"</span>, <span class="st">"BP"</span><span class="op">=</span><span class="st">"BP"</span><span class="op">)</span>,</span>
<span>                                 methods<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slopehunter"</span>,<span class="st">"mr_collider_bias"</span>,<span class="st">"dudbridge"</span><span class="op">)</span>,</span>
<span>                                 tsmr_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mr_ivw"</span>,<span class="st">"mr_egger_regression"</span>,<span class="st">"mr_simple_median"</span>,<span class="st">"mr_simple_mode"</span><span class="op">)</span>,</span>
<span>                                 ip     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>,<span class="fl">0.001</span>,<span class="fl">0.0001</span>,<span class="fl">0.00001</span>,<span class="fl">0.000001</span>,<span class="fl">0.0000001</span><span class="op">)</span>,</span>
<span>                                 pi0    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span><span class="op">)</span>,</span>
<span>                                 sxy1   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-05</span><span class="op">)</span>,</span>
<span>                                 bootstraps <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_slope.html">plot_slope</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/collider_slopes.png" width="98%" style="display: block; margin: auto;"></p>
<p>Plotted another way with
<code><a href="../reference/plot_correction_stability.html">plot_correction_stability()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_correction_stability.html">plot_correction_stability</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/collider_stability.png" width="98%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="applying-the-correction-factor">Applying the correction factor<a class="anchor" aria-label="anchor" href="#applying-the-correction-factor"></a>
</h2>
<p>Once the parameters have been set and <code>b</code> estimate of the
correction calculated we next need to apply the correction to the
(biased) progression GWAS data. As the correction uses the incidence
data we can only apply this to variants that appear in both incidence
and progression datasets (the harmonised dataset).</p>
<p><span class="math display">\[
\beta_{GP} = \beta_{GP}^{'} - b_{1}\beta_{GI}
\]</span></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply correction</span></span>
<span><span class="va">corrected_gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apply_collider_correction.html">apply_collider_correction</a></span><span class="op">(</span>gwas_i <span class="op">=</span> <span class="va">gwas_incidence</span>, </span>
<span>                                            gwas_p <span class="op">=</span> <span class="va">gwas_progression</span>,</span>
<span>                                            merge  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span><span class="op">=</span><span class="st">"CHR"</span>, <span class="st">"BP"</span><span class="op">=</span><span class="st">"BP"</span><span class="op">)</span>,</span>
<span>                                            b_correction_factor <span class="op">=</span> <span class="fl">0.6473</span>,</span>
<span>                                            b_std_err <span class="op">=</span> <span class="fl">0.2162</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate difference in raw and adjusted betas; and effect on genome wide significance</span></span>
<span><span class="va">corrected_gwas</span><span class="op">[</span>, <span class="va">beta_diff</span> <span class="op">:=</span> <span class="va">BETA_progression</span> <span class="op">-</span> <span class="va">adjusted_beta</span><span class="op">]</span></span>
<span><span class="va">sig_change_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dropped hit"</span>, <span class="st">"New hit"</span>, <span class="st">"No change"</span><span class="op">)</span></span>
<span><span class="va">corrected_gwas</span><span class="op">[</span>, <span class="va">sig_change</span><span class="op">:=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fcase.html" class="external-link">fcase</a></span><span class="op">(</span><span class="va">P_progression</span> <span class="op">&lt;</span> <span class="fl">5e-8</span> <span class="op">&amp;</span> <span class="va">adjusted_p</span> <span class="op">&gt;=</span> <span class="fl">5e-8</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="st">"Dropped hit"</span>, levels<span class="op">=</span><span class="va">sig_change_levels</span><span class="op">)</span>,</span>
<span>                                                <span class="va">P_progression</span> <span class="op">&gt;=</span><span class="fl">5e-8</span> <span class="op">&amp;</span> <span class="va">adjusted_p</span> <span class="op">&lt;</span>  <span class="fl">5e-8</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="st">"New hit"</span>, levels<span class="op">=</span><span class="va">sig_change_levels</span><span class="op">)</span>,</span>
<span>                                                default <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="st">"No change"</span>, levels<span class="op">=</span><span class="va">sig_change_levels</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># plot difference</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data    <span class="op">=</span> <span class="va">corrected_gwas</span>,</span>
<span>                     mapping <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">beta_diff</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">sig_change</span>, </span>
<span>                                             fill  <span class="op">=</span> <span class="va">adjusted_beta</span>, </span>
<span>                                             color <span class="op">=</span> <span class="va">sig_change</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.95</span>, position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_jitter.html" class="external-link">position_jitter</a></span><span class="op">(</span>seed <span class="op">=</span> <span class="fl">1</span>, width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Change in genome wide significance"</span>, </span>
<span>                              values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dropped hit"</span><span class="op">=</span><span class="fl">25</span>, <span class="st">"New hit"</span><span class="op">=</span><span class="fl">24</span>, <span class="st">"No change"</span><span class="op">=</span><span class="fl">21</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Change in genome wide significance"</span>,</span>
<span>                              values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dropped hit"</span><span class="op">=</span><span class="st">"red"</span>, <span class="st">"New hit"</span><span class="op">=</span><span class="st">"green"</span>, <span class="st">"No change"</span><span class="op">=</span><span class="st">"#44015410"</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span>, fill<span class="op">=</span><span class="st">"purple"</span>, color<span class="op">=</span><span class="st">"#44015410"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">'Slope-hunter adjustment'</span>,</span>
<span>                x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"\u03B2"</span><span class="op">[</span><span class="va">progression</span><span class="op">-</span><span class="va">SNPs</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"\u03B2"</span><span class="op">[</span><span class="va">adjustment</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                color <span class="op">=</span> <span class="st">"Change in genome wide significance"</span>, </span>
<span>                fill <span class="op">=</span> <span class="st">"BETA magnitude"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/collider_adjustment.png" width="98%" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
