<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LD matrix â€¢ genepi.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LD matrix">
<meta property="og:description" content="genepi.utils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LD matrix</h1>
            
      
      
      <div class="hidden name"><code>ld_matrix.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>See the setup process for the PLINK2 executable and reference file in
the <a href="https://nicksunderland.github.io/genepi.utils/articles/clumping.html" class="external-link">Clumping</a>
vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genepi.utils</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the gwas data</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>dat<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example2_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span>, map<span class="op">=</span><span class="st">"ns_map"</span>, fill_rsid<span class="op">=</span><span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading GWAS ...</span></span>
<span><span class="co">#&gt; [i] loading data from: example2_gwas_sumstats.tsv</span></span>
<span><span class="co">#&gt; [i] applying column mapping</span></span>
<span><span class="co">#&gt; [i] enforcing column types</span></span>
<span><span class="co">#&gt; [i] standardising columns</span></span>
<span><span class="co">#&gt; [i] standardising allele coding</span></span>
<span><span class="co">#&gt; [i] applying filters</span></span>
<span><span class="co">#&gt; [i] checking rsid coding</span></span>
<span><span class="co">#&gt;  [?] 10 RSIDs could not be parsed, attempting to fetch from dbSNP.</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (10/10), of which 0% (0/10) where found flipping alleles</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.data.table.html">as.data.table</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check path to reference</span></span>
<span><span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span><span class="st">"GRCh37"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/xx20081/Documents/local_data/genome_reference/ref_1000GP_phase3_plink/EUR_1kGv3"</span></span>
<span></span>
<span><span class="co"># check path to plink2 executable</span></span>
<span><span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/plink/plink2"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ld-matrix">LD matrix<a class="anchor" aria-label="anchor" href="#ld-matrix"></a>
</h2>
<p>Generate the LD matrix. If we provide a GWAS data.table with the
alleles then it will be harmonised with the LD matrix.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># produce LD matrix </span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span><span class="va">gwas</span>,</span>
<span>                 method    <span class="op">=</span> <span class="st">"r2"</span>,</span>
<span>                 plink2    <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                 plink_ref <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span>build<span class="op">=</span><span class="st">"GRCh37"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; ld_matrix(): variant alleles provided, returning harmonised data and LD matrix object</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="harmonise-data-against-ld-matrix-alleles">Harmonise data against LD matrix alleles<a class="anchor" aria-label="anchor" href="#harmonise-data-against-ld-matrix-alleles"></a>
</h2>
<p>The previous command has harmonised the <code>gwas</code> input
data.table to the LD alleles. The return value is a list with elements
<code>dat</code> and <code>ld_mat</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># harmonised data </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">obj</span><span class="op">[</span><span class="st">'dat'</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $dat</span></span>
<span><span class="co">#&gt;           rsid    chr        bp     ea     oa       eaf   beta    se     p</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;char&gt;     &lt;int&gt; &lt;char&gt; &lt;char&gt;     &lt;num&gt;  &lt;num&gt; &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:  rs7899632     10 100000625      A      G 0.9999290  0.037 0.137 0.781</span></span>
<span><span class="co">#&gt;  2: rs61875309     10 100000645      A      C 0.8790000 -0.045 0.173 0.961</span></span>
<span><span class="co">#&gt;  3: rs12258651     10 100003242      T      G 0.9999995  0.037 0.134 0.480</span></span>
<span><span class="co">#&gt;  4:  rs1359508     10 100003785      T      C 0.8450000  0.008 0.137 0.743</span></span>
<span><span class="co">#&gt;  5:  rs1048754     10 100004360      G      A 0.1620000 -0.045 0.134 0.961</span></span>
<span><span class="co">#&gt;  6:  rs3750595     10 100004906      C      A 0.2600000  0.037 0.173 0.781</span></span>
<span><span class="co">#&gt;  7:  rs2025625     10 100004996      G      A 1.0000000  0.008 0.137 0.743</span></span>
<span><span class="co">#&gt;  8: rs10786405     10 100005282      C      T 0.9990700  0.147 0.134 0.781</span></span>
<span><span class="co">#&gt;  9: rs11816998     10 100007362      G      C 0.9999184 -0.045 0.173 0.961</span></span>
<span><span class="co">#&gt; 10:  rs3793692     10 100008436      G      A 0.9990540  0.008 0.208 0.743</span></span>
<span><span class="co">#&gt;         n ncase  trait     id LD_ref LD_alt LD_rsid_allele</span></span>
<span><span class="co">#&gt;     &lt;int&gt; &lt;int&gt; &lt;char&gt; &lt;char&gt; &lt;char&gt; &lt;char&gt;         &lt;char&gt;</span></span>
<span><span class="co">#&gt;  1:    NA    NA  trait     id      A      G  rs7899632_A_G</span></span>
<span><span class="co">#&gt;  2:    NA    NA  trait     id      A      C rs61875309_A_C</span></span>
<span><span class="co">#&gt;  3:    NA    NA  trait     id      T      G rs12258651_T_G</span></span>
<span><span class="co">#&gt;  4:    NA    NA  trait     id      T      C  rs1359508_T_C</span></span>
<span><span class="co">#&gt;  5:    NA    NA  trait     id      G      A  rs1048754_G_A</span></span>
<span><span class="co">#&gt;  6:    NA    NA  trait     id      C      A  rs3750595_C_A</span></span>
<span><span class="co">#&gt;  7:    NA    NA  trait     id      G      A  rs2025625_G_A</span></span>
<span><span class="co">#&gt;  8:    NA    NA  trait     id      C      T rs10786405_C_T</span></span>
<span><span class="co">#&gt;  9:    NA    NA  trait     id      G      C rs11816998_G_C</span></span>
<span><span class="co">#&gt; 10:    NA    NA  trait     id      G      A  rs3793692_G_A</span></span>
<span></span>
<span><span class="co"># the LD matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">obj</span><span class="op">[</span><span class="st">'ld_mat'</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $ld_mat</span></span>
<span><span class="co">#&gt;                rs7899632_A_G rs61875309_A_C rs12258651_T_G rs1359508_T_C</span></span>
<span><span class="co">#&gt; rs7899632_A_G      1.0000000     0.23544930     0.20391148    0.38781072</span></span>
<span><span class="co">#&gt; rs61875309_A_C     0.2354493     1.00000000     0.04801081    0.13827704</span></span>
<span><span class="co">#&gt; rs12258651_T_G     0.2039115     0.04801081     1.00000000    0.07907906</span></span>
<span><span class="co">#&gt; rs1359508_T_C      0.3878107     0.13827704     0.07907906    1.00000000</span></span>
<span><span class="co">#&gt; rs1048754_G_A      0.2354493     1.00000000     0.04801081    0.13827704</span></span>
<span><span class="co">#&gt; rs3750595_C_A      1.0000000     0.23544930     0.20391148    0.38781072</span></span>
<span><span class="co">#&gt; rs2025625_G_A      0.3878107     0.13827704     0.07907906    1.00000000</span></span>
<span><span class="co">#&gt; rs10786405_C_T     1.0000000     0.23544930     0.20391148    0.38781072</span></span>
<span><span class="co">#&gt; rs11816998_G_C     0.2314337     0.98294511     0.04719200    0.13591874</span></span>
<span><span class="co">#&gt; rs3793692_G_A      0.3878107     0.13827704     0.07907906    1.00000000</span></span>
<span><span class="co">#&gt;                rs1048754_G_A rs3750595_C_A rs2025625_G_A rs10786405_C_T</span></span>
<span><span class="co">#&gt; rs7899632_A_G     0.23544930     1.0000000    0.38781072      1.0000000</span></span>
<span><span class="co">#&gt; rs61875309_A_C    1.00000000     0.2354493    0.13827704      0.2354493</span></span>
<span><span class="co">#&gt; rs12258651_T_G    0.04801081     0.2039115    0.07907906      0.2039115</span></span>
<span><span class="co">#&gt; rs1359508_T_C     0.13827704     0.3878107    1.00000000      0.3878107</span></span>
<span><span class="co">#&gt; rs1048754_G_A     1.00000000     0.2354493    0.13827704      0.2354493</span></span>
<span><span class="co">#&gt; rs3750595_C_A     0.23544930     1.0000000    0.38781072      1.0000000</span></span>
<span><span class="co">#&gt; rs2025625_G_A     0.13827704     0.3878107    1.00000000      0.3878107</span></span>
<span><span class="co">#&gt; rs10786405_C_T    0.23544930     1.0000000    0.38781072      1.0000000</span></span>
<span><span class="co">#&gt; rs11816998_G_C    0.98294511     0.2314337    0.13591874      0.2314337</span></span>
<span><span class="co">#&gt; rs3793692_G_A     0.13827704     0.3878107    1.00000000      0.3878107</span></span>
<span><span class="co">#&gt;                rs11816998_G_C rs3793692_G_A</span></span>
<span><span class="co">#&gt; rs7899632_A_G       0.2314337    0.38781072</span></span>
<span><span class="co">#&gt; rs61875309_A_C      0.9829451    0.13827704</span></span>
<span><span class="co">#&gt; rs12258651_T_G      0.0471920    0.07907906</span></span>
<span><span class="co">#&gt; rs1359508_T_C       0.1359187    1.00000000</span></span>
<span><span class="co">#&gt; rs1048754_G_A       0.9829451    0.13827704</span></span>
<span><span class="co">#&gt; rs3750595_C_A       0.2314337    0.38781072</span></span>
<span><span class="co">#&gt; rs2025625_G_A       0.1359187    1.00000000</span></span>
<span><span class="co">#&gt; rs10786405_C_T      0.2314337    0.38781072</span></span>
<span><span class="co">#&gt; rs11816998_G_C      1.0000000    0.13591874</span></span>
<span><span class="co">#&gt; rs3793692_G_A       0.1359187    1.00000000</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
