<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LD matrix â€¢ genepi.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LD matrix">
<meta property="og:description" content="genepi.utils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LD matrix</h1>
            
      
      
      <div class="hidden name"><code>ld_matrix.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>See the setup process for the PLINK2 executable and reference file in
the <a href="https://nicksunderland.github.io/genepi.utils/articles/clumping.html" class="external-link">Clumping</a>
vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genepi.utils</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the gwas data</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example2_gwas_sumstats.tsv"</span>, package<span class="op">=</span><span class="st">"genepi.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ensure annotated with rsID</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/standardise_gwas.html">standardise_gwas</a></span><span class="op">(</span><span class="va">gwas</span>, input_format<span class="op">=</span><span class="st">"ns_map"</span>, build<span class="op">=</span><span class="st">"GRCh37"</span>, populate_rsid<span class="op">=</span><span class="st">"b37_dbsnp156"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Standardising GWAS ...</span></span>
<span><span class="co">#&gt; RSID mapping...</span></span>
<span><span class="co">#&gt; Chromosomes to process:  10</span></span>
<span><span class="co">#&gt; RSID coverage 100% (10/10), of which 0% (0/10) where found flipping alleles</span></span>
<span></span>
<span><span class="co"># check path to reference</span></span>
<span><span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span><span class="st">"GRCh37"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/xx20081/Documents/local_data/genome_reference/ref_1000GP_phase3_plink/all_phase3_nodup"</span></span>
<span></span>
<span><span class="co"># check path to plink2 executable</span></span>
<span><span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/plink/plink2"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ld-matrix">LD matrix<a class="anchor" aria-label="anchor" href="#ld-matrix"></a>
</h2>
<p>Generate the LD matrix.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># produce LD matrix </span></span>
<span><span class="va">ld_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, <span class="st">"RSID"</span><span class="op">]</span>,</span>
<span>                    plink2    <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                    plink_ref <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="../reference/which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span>build<span class="op">=</span><span class="st">"GRCh37"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># view result</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ld_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt;            rs7899632 rs61875309 rs12258651 rs1359508 rs1048754 rs3750595</span></span>
<span><span class="co">#&gt; rs7899632   1.000000  -0.352729   0.464422 -0.605224 -0.351842  0.999186</span></span>
<span><span class="co">#&gt; rs61875309 -0.352729   1.000000  -0.163815 -0.273537  0.997485 -0.352442</span></span>
<span><span class="co">#&gt; rs12258651  0.464422  -0.163815   1.000000 -0.281882 -0.163403  0.463646</span></span>
<span><span class="co">#&gt; rs1359508  -0.605224  -0.273537  -0.281882  1.000000 -0.276556 -0.605593</span></span>
<span><span class="co">#&gt; rs1048754  -0.351842   0.997485  -0.163403 -0.276556  1.000000 -0.351555</span></span>
<span><span class="co">#&gt; rs3750595   0.999186  -0.352442   0.463646 -0.605593 -0.351555  1.000000</span></span>
<span><span class="co">#&gt; rs2025625  -0.602476  -0.275604  -0.280205  0.994053 -0.274911 -0.602850</span></span>
<span><span class="co">#&gt; rs10786405  0.989851  -0.349437   0.467642 -0.599557 -0.348558  0.990658</span></span>
<span><span class="co">#&gt; rs11816998 -0.346224   0.984860  -0.158549 -0.270245  0.982267 -0.345940</span></span>
<span><span class="co">#&gt;            rs2025625 rs10786405 rs11816998</span></span>
<span><span class="co">#&gt; rs7899632  -0.602476   0.989851  -0.346224</span></span>
<span><span class="co">#&gt; rs61875309 -0.275604  -0.349437   0.984860</span></span>
<span><span class="co">#&gt; rs12258651 -0.280205   0.467642  -0.158549</span></span>
<span><span class="co">#&gt; rs1359508   0.994053  -0.599557  -0.270245</span></span>
<span><span class="co">#&gt; rs1048754  -0.274911  -0.348558   0.982267</span></span>
<span><span class="co">#&gt; rs3750595  -0.602850   0.990658  -0.345940</span></span>
<span><span class="co">#&gt; rs2025625   1.000000  -0.596843  -0.272363</span></span>
<span><span class="co">#&gt; rs10786405 -0.596843   1.000000  -0.342968</span></span>
<span><span class="co">#&gt; rs11816998 -0.272363  -0.342968   1.000000</span></span>
<span><span class="co">#&gt; attr(,"log")</span></span>
<span><span class="co">#&gt; [1] "PLINK v2.00a6 M1 (23 Nov 2023)\nOptions in effect:\n  --extract /var/folders/62/n2yrhw752hn73nlbzp0r0_rr0000gq/T//RtmpMOBawz/file124e549dc95bd\n  --out /Users/xx20081/Downloads/ld_mat\n  --pfile /Users/xx20081/Documents/local_data/genome_reference/ref_1000GP_phase3_plink/all_phase3_nodup\n  --r-phased square\n\nHostname: K3P0KGKVKT\nWorking directory: /Users/xx20081/git/genepi.utils/vignettes\nStart time: Tue Dec  5 20:08:58 2023\n\nRandom number seed: 1701806938\n98304 MiB RAM detected; reserving 49152 MiB for main workspace.\nUsing up to 12 threads (change this with --threads).\n2504 samples (1270 females, 1234 males; 2497 founders) loaded from\n/Users/xx20081/Documents/local_data/genome_reference/ref_1000GP_phase3_plink/all_phase3_nodup.psam.\n48493710 variants loaded from\n/Users/xx20081/Documents/local_data/genome_reference/ref_1000GP_phase3_plink/all_phase3_nodup.pvar.\n2 categorical phenotypes loaded.\n--extract: 9 variants remaining.\nCalculating allele frequencies... done.\n9 variants remaining after main filters.\n--r-phased: Variant IDs written to\n/Users/xx20081/Downloads/ld_mat.phased.vcor1.vars .\n--r-phased: Matrix written to /Users/xx20081/Downloads/ld_mat.phased.vcor1 .\n\nEnd time: Tue Dec  5 20:09:03 2023"</span></span>
<span><span class="co">#&gt; attr(,"allele_info")</span></span>
<span><span class="co">#&gt;         RSID CHR        BP REF ALT</span></span>
<span><span class="co">#&gt; 1  rs7899632  10 100000625   A   G</span></span>
<span><span class="co">#&gt; 2 rs61875309  10 100000645   A   C</span></span>
<span><span class="co">#&gt; 3 rs12258651  10 100003242   T   G</span></span>
<span><span class="co">#&gt; 4  rs1359508  10 100003785   T   C</span></span>
<span><span class="co">#&gt; 5  rs1048754  10 100004360   G   A</span></span>
<span><span class="co">#&gt; 6  rs3750595  10 100004906   C   A</span></span>
<span><span class="co">#&gt; 7  rs2025625  10 100004996   G   A</span></span>
<span><span class="co">#&gt; 8 rs10786405  10 100005282   C   T</span></span>
<span><span class="co">#&gt; 9 rs11816998  10 100007362   G   C</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="harmonise-data-against-ld-matrix-alleles">Harmonise data against LD matrix alleles<a class="anchor" aria-label="anchor" href="#harmonise-data-against-ld-matrix-alleles"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># harmonised data </span></span>
<span><span class="va">harm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/harmonise.html">harmonise</a></span><span class="op">(</span><span class="va">gwas</span>, <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/copy.html" class="external-link">copy</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span>, gwas1_trait<span class="op">=</span><span class="st">"exposure"</span>, gwas2_trait<span class="op">=</span><span class="st">"outcome"</span>, merge<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RSID"</span><span class="op">=</span><span class="st">"RSID"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mess up the alleles vs reference</span></span>
<span><span class="va">harm</span> <span class="op">&lt;-</span> <span class="va">harm</span><span class="op">[</span><span class="va">RSID_exposure</span><span class="op">==</span><span class="st">"rs7899632"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EA_exposure"</span>,<span class="st">"EA_outcome"</span>,<span class="st">"OA_exposure"</span>,<span class="st">"OA_outcome"</span><span class="op">)</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"A"</span>,<span class="st">"G"</span>,<span class="st">"G"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">harm</span><span class="op">[</span><span class="va">RSID_exposure</span><span class="op">==</span><span class="st">"rs7899632"</span>, <span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="co">#&gt; Key: &lt;RSID_exposure&gt;</span></span>
<span><span class="co">#&gt;            SNP_exposure          SNP_outcome CHR_exposure CHR_outcome</span></span>
<span><span class="co">#&gt;                  &lt;char&gt;               &lt;char&gt;       &lt;char&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: 10:100000625[b37]A,G 10:100000625[b37]A,G           10          10</span></span>
<span><span class="co">#&gt;    BP_exposure BP_outcome EA_exposure EA_outcome OA_exposure OA_outcome</span></span>
<span><span class="co">#&gt;          &lt;int&gt;      &lt;int&gt;      &lt;char&gt;     &lt;char&gt;      &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:   100000625  100000625           A          A           G          G</span></span>
<span><span class="co">#&gt;    EAF_exposure EAF_outcome BETA_exposure BETA_outcome P_exposure P_outcome</span></span>
<span><span class="co">#&gt;           &lt;num&gt;       &lt;num&gt;         &lt;num&gt;        &lt;num&gt;      &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:      7.1e-05     7.1e-05        -0.037       -0.037      0.781     0.781</span></span>
<span><span class="co">#&gt;    palindromic   keep RSID_exposure SE_exposure SE_outcome RSID_outcome</span></span>
<span><span class="co">#&gt;         &lt;lgcl&gt; &lt;lgcl&gt;        &lt;char&gt;       &lt;num&gt;      &lt;num&gt;       &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:       FALSE   TRUE     rs7899632       0.137      0.137    rs7899632</span></span>
<span></span>
<span><span class="co"># re-harmonise</span></span>
<span><span class="va">harm_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/harmonise_ld_dat.html">harmonise_ld_dat</a></span><span class="op">(</span><span class="va">harm</span>, <span class="va">ld_mat</span>, gwas1_trait<span class="op">=</span><span class="st">"exposure"</span>, gwas2_trait<span class="op">=</span><span class="st">"outcome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># harmonised data</span></span>
<span><span class="va">harm_result</span><span class="op">[[</span><span class="st">"harm"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">RSID_exposure</span><span class="op">==</span><span class="st">"rs7899632"</span>, <span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="co">#&gt;            SNP_exposure          SNP_outcome CHR_exposure CHR_outcome</span></span>
<span><span class="co">#&gt;                  &lt;char&gt;               &lt;char&gt;       &lt;char&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: 10:100000625[b37]A,G 10:100000625[b37]A,G           10          10</span></span>
<span><span class="co">#&gt;    BP_exposure BP_outcome EA_exposure EA_outcome OA_exposure OA_outcome</span></span>
<span><span class="co">#&gt;          &lt;int&gt;      &lt;int&gt;      &lt;char&gt;     &lt;char&gt;      &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:   100000625  100000625           G          G           A          A</span></span>
<span><span class="co">#&gt;    EAF_exposure EAF_outcome BETA_exposure BETA_outcome P_exposure P_outcome</span></span>
<span><span class="co">#&gt;           &lt;num&gt;       &lt;num&gt;         &lt;num&gt;        &lt;num&gt;      &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:     0.999929    0.999929         0.037        0.037      0.781     0.781</span></span>
<span><span class="co">#&gt;    palindromic   keep RSID_exposure SE_exposure SE_outcome RSID_outcome CHR_LD</span></span>
<span><span class="co">#&gt;         &lt;lgcl&gt; &lt;lgcl&gt;        &lt;char&gt;       &lt;num&gt;      &lt;num&gt;       &lt;char&gt;  &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:       FALSE   TRUE     rs7899632       0.137      0.137    rs7899632     10</span></span>
<span><span class="co">#&gt;        BP_LD REF_LD ALT_LD keep_ld</span></span>
<span><span class="co">#&gt;        &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt; 1: 100000625      A      G    TRUE</span></span>
<span></span>
<span><span class="co"># LD matrix</span></span>
<span><span class="va">harm_result</span><span class="op">[[</span><span class="st">"ld_mat"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;            rs7899632 rs61875309 rs12258651 rs1359508 rs1048754 rs3750595</span></span>
<span><span class="co">#&gt; rs7899632   1.000000  -0.352729   0.464422 -0.605224 -0.351842  0.999186</span></span>
<span><span class="co">#&gt; rs61875309 -0.352729   1.000000  -0.163815 -0.273537  0.997485 -0.352442</span></span>
<span><span class="co">#&gt; rs12258651  0.464422  -0.163815   1.000000 -0.281882 -0.163403  0.463646</span></span>
<span><span class="co">#&gt; rs1359508  -0.605224  -0.273537  -0.281882  1.000000 -0.276556 -0.605593</span></span>
<span><span class="co">#&gt; rs1048754  -0.351842   0.997485  -0.163403 -0.276556  1.000000 -0.351555</span></span>
<span><span class="co">#&gt; rs3750595   0.999186  -0.352442   0.463646 -0.605593 -0.351555  1.000000</span></span>
<span><span class="co">#&gt; rs2025625  -0.602476  -0.275604  -0.280205  0.994053 -0.274911 -0.602850</span></span>
<span><span class="co">#&gt; rs10786405  0.989851  -0.349437   0.467642 -0.599557 -0.348558  0.990658</span></span>
<span><span class="co">#&gt; rs11816998 -0.346224   0.984860  -0.158549 -0.270245  0.982267 -0.345940</span></span>
<span><span class="co">#&gt;            rs2025625 rs10786405 rs11816998</span></span>
<span><span class="co">#&gt; rs7899632  -0.602476   0.989851  -0.346224</span></span>
<span><span class="co">#&gt; rs61875309 -0.275604  -0.349437   0.984860</span></span>
<span><span class="co">#&gt; rs12258651 -0.280205   0.467642  -0.158549</span></span>
<span><span class="co">#&gt; rs1359508   0.994053  -0.599557  -0.270245</span></span>
<span><span class="co">#&gt; rs1048754  -0.274911  -0.348558   0.982267</span></span>
<span><span class="co">#&gt; rs3750595  -0.602850   0.990658  -0.345940</span></span>
<span><span class="co">#&gt; rs2025625   1.000000  -0.596843  -0.272363</span></span>
<span><span class="co">#&gt; rs10786405 -0.596843   1.000000  -0.342968</span></span>
<span><span class="co">#&gt; rs11816998 -0.272363  -0.342968   1.000000</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nicholas Sunderland, Andrew Elmore.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
