<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Clump a GWAS — clump • genepi.utils</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Clump a GWAS — clump"><meta name="description" content="Clump variants in a GWAS using PLINK2 and an appropriate reference panel.
For example, the 1000 genomes phase 3 data can be downloaded from the PLINK
website (https://www.cog-genomics.org/plink/2.0/resources#phase3_1kg). To remove
duplicates you can run:

plink2
–pfile all_phase3
–rm-dup force-first
–make-pgen
–out all_phase3_nodup

The path to the reference (without the plink extensions) should be passed as the
plink_ref argument. The path to the plink2 executable should be passed as the
plink2 argument."><meta property="og:description" content="Clump variants in a GWAS using PLINK2 and an appropriate reference panel.
For example, the 1000 genomes phase 3 data can be downloaded from the PLINK
website (https://www.cog-genomics.org/plink/2.0/resources#phase3_1kg). To remove
duplicates you can run:

plink2
–pfile all_phase3
–rm-dup force-first
–make-pgen
–out all_phase3_nodup

The path to the reference (without the plink extensions) should be passed as the
plink_ref argument. The path to the plink2 executable should be passed as the
plink2 argument."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genepi.utils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a></li>
    <li><a class="dropdown-item" href="../articles/clumping.html">Clumping</a></li>
    <li><a class="dropdown-item" href="../articles/collider_bias.html">Collider bias</a></li>
    <li><a class="dropdown-item" href="../articles/ld_matrix.html">LD matrix</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Plotting</a></li>
    <li><a class="dropdown-item" href="../articles/rationale.html">Rationale and key points</a></li>
    <li><a class="dropdown-item" href="../articles/standardise_gwas.html">Standardise GWAS</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Clump a GWAS</h1>

      <div class="d-none name"><code>clump.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Clump variants in a GWAS using PLINK2 and an appropriate reference panel.
For example, the 1000 genomes phase 3 data can be downloaded from the PLINK
website (https://www.cog-genomics.org/plink/2.0/resources#phase3_1kg). To remove
duplicates you can run:
<br><br>
plink2 <br>
–pfile all_phase3 <br>
–rm-dup force-first <br>
–make-pgen <br>
–out all_phase3_nodup <br><br>
The path to the reference (without the plink extensions) should be passed as the
<code>plink_ref</code> argument. The path to the plink2 executable should be passed as the
<code>plink2</code> argument.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clump</span><span class="op">(</span></span>
<span>  <span class="va">gwas</span>,</span>
<span>  p1 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  p2 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  r2 <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  kb <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  plink2 <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="which_plink2.html">which_plink2</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plink_ref <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="which_1000G_reference.html">which_1000G_reference</a></span><span class="op">(</span>build <span class="op">=</span> <span class="st">"GRCh37"</span><span class="op">)</span>,</span>
<span>  logging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel_cores <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gwas">gwas<a class="anchor" aria-label="anchor" href="#arg-gwas"></a></dt>
<dd><p>a data.frame like object with at least columns rsid, ea, oa, and p</p></dd>


<dt id="arg-p-">p1<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>a numeric, the p-value threshold for inclusion as a clump</p></dd>


<dt id="arg-p-">p2<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>a numeric, the p-value threshold for incorporation into a clump</p></dd>


<dt id="arg-r-">r2<a class="anchor" aria-label="anchor" href="#arg-r-"></a></dt>
<dd><p>a numeric, the r2 value</p></dd>


<dt id="arg-kb">kb<a class="anchor" aria-label="anchor" href="#arg-kb"></a></dt>
<dd><p>a integer, the window for clumping</p></dd>


<dt id="arg-plink-">plink2<a class="anchor" aria-label="anchor" href="#arg-plink-"></a></dt>
<dd><p>a string, path to the plink executable</p></dd>


<dt id="arg-plink-ref">plink_ref<a class="anchor" aria-label="anchor" href="#arg-plink-ref"></a></dt>
<dd><p>a string, path to the pfile genome reference</p></dd>


<dt id="arg-logging">logging<a class="anchor" aria-label="anchor" href="#arg-logging"></a></dt>
<dd><p>a logical, whether to set the plink logging information as attributes (<code>log</code>, <code>missing_id</code>, <code>missing_allele</code>) on the returned data.table</p></dd>


<dt id="arg-parallel-cores">parallel_cores<a class="anchor" aria-label="anchor" href="#arg-parallel-cores"></a></dt>
<dd><p>an integer, how many cores / threads to use</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a data.table with additional columns <code>index</code> (logical, whether the variant is an index SNP) and <code>clump</code> (integer, the clump the variant belongs to)</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

