<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>GWAS object — GWAS • genepi.utils</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS object — GWAS"><meta property="og:description" content="A GWAS object is a container for vectors of GWAS data, a correlation matrix, and
meta-data regarding quality control procedures applied at the point of object creation / data import."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genepi.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a>
    </li>
    <li>
      <a href="../articles/clumping.html">Clumping</a>
    </li>
    <li>
      <a href="../articles/collider_bias.html">Collider bias</a>
    </li>
    <li>
      <a href="../articles/drug_target_proxy.html">Drug target proxies</a>
    </li>
    <li>
      <a href="../articles/harmonise.html">Harmonising datasets</a>
    </li>
    <li>
      <a href="../articles/ld_matrix.html">LD matrix</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/rationale.html">Rationale and key points</a>
    </li>
    <li>
      <a href="../articles/standardise_gwas.html">Standardise GWAS</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GWAS object</h1>
    
    <div class="hidden name"><code>GWAS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A GWAS object is a container for vectors of GWAS data, a correlation matrix, and
meta-data regarding quality control procedures applied at the point of object creation / data import.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GWAS</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  map <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  drop <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fill <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fill_rsid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  missing_rsid <span class="op">=</span> <span class="st">"fill_CHR:BP"</span>,</span>
<span>  parallel_cores <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  dbsnp_dir <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="which_dbsnp_directory.html">which_dbsnp_directory</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>beta_invalid <span class="op">=</span> <span class="st">"!is.infinite(beta) &amp; abs(beta) &lt; 20"</span>, eaf_invalid <span class="op">=</span></span>
<span>    <span class="st">"eaf &gt; 0 &amp; eaf &lt; 1"</span>, p_invalid <span class="op">=</span> <span class="st">"!is.infinite(p)"</span>, se_invalid <span class="op">=</span> <span class="st">"!is.infinite(se)"</span>,</span>
<span>    alleles_invalid <span class="op">=</span> <span class="st">"!is.na(ea) &amp; !is.na(oa)"</span>, chr_missing <span class="op">=</span> <span class="st">"!is.na(chr)"</span>, bp_missing</span>
<span>    <span class="op">=</span> <span class="st">"!is.na(bp)"</span>, beta_missing <span class="op">=</span> <span class="st">"!is.na(beta)"</span>, se_missing <span class="op">=</span> <span class="st">"!is.na(se)"</span>, p_missing <span class="op">=</span></span>
<span>    <span class="st">"!is.na(p)"</span>, eaf_missing <span class="op">=</span> <span class="st">"!is.na(eaf)"</span><span class="op">)</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_map <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dat</dt>
<dd><p>a valid string file path to be read by <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> or a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table::data.table</a></code> object; the GWAS data source</p></dd>


<dt>map</dt>
<dd><p>a valid input to the <code>ColumnMap</code> class constructor (a predefined map string id, a named list or character vector, or a ColumMap object)</p></dd>


<dt>drop</dt>
<dd><p>a logical, whether to drop data source columns not in the column <code>map</code></p></dd>


<dt>fill</dt>
<dd><p>a logical, whether to add (NAs) missing columns present in the column <code>map</code> but not present in the data source</p></dd>


<dt>fill_rsid</dt>
<dd><p>either FALSE or a valid argument for the chrpos_to_rsid <code>build</code> argument, e.g. "b37_dbsnp156"</p></dd>


<dt>missing_rsid</dt>
<dd><p>a string, how to handle missing rsids: one of "fill_CHR:BP", "fill_CHR:BP_OA_EA", "overwrite_CHR:BP",
"overwrite_CHR:BP:OA:EA", "none", or "leave"</p></dd>


<dt>parallel_cores</dt>
<dd><p>an integer, number of cores to used for RSID mapping, default is maximum machine cores</p></dd>


<dt>dbsnp_dir</dt>
<dd><p>path to the dbsnp directory of fst files see chrpos_to_rsid <code>dbsnp_dir</code> argument</p></dd>


<dt>filters</dt>
<dd><p>a list of named strings, each to be evaluated as an expression to filter the data during the quality control steps (above)</p></dd>


<dt>reference</dt>
<dd><p>a valid string file path to be read by <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> or a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table::data.table</a></code> object; the reference data</p></dd>


<dt>ref_map</dt>
<dd><p>a valid input to the <code>ColumnMap</code> class constructor (a predefined map id (a string), a named list or character vector, or a
ColumMap object) defining at least columns <code>rsid</code> (or <code>chr</code>, <code>bp</code>), <code>ea</code>, <code>oa</code> and <code>eaf</code>.</p></dd>


<dt>verbose</dt>
<dd><p>a logical, whether to print details</p></dd>


<dt>...</dt>
<dd><p>variable capture to be passed to the constructor, e.g. individual vectors for the slots, rather that <code>dat</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>an S7 class genepi.utils::GWAS object</p>
    </div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt><code>rsid</code></dt>
<dd><p>character, variant ID - usually in rs12345 format, however this can be changed with the <code>missing_rsid</code> argument</p></dd>


<dt><code>chr</code></dt>
<dd><p>character, chromosome identifier</p></dd>


<dt><code>bp</code></dt>
<dd><p>integer, base position</p></dd>


<dt><code>ea</code></dt>
<dd><p>character, effect allele</p></dd>


<dt><code>oa</code></dt>
<dd><p>character, other allele</p></dd>


<dt><code>eaf</code></dt>
<dd><p>numeric, effect allele frequency</p></dd>


<dt><code>beta</code></dt>
<dd><p>numeric, effect size</p></dd>


<dt><code>se</code></dt>
<dd><p>numeric, effect size standard error</p></dd>


<dt><code>p</code></dt>
<dd><p>numeric, p-value</p></dd>


<dt><code>n</code></dt>
<dd><p>integer, total number of samples</p></dd>


<dt><code>ncase</code></dt>
<dd><p>integer, number of cases</p></dd>


<dt><code>strand</code></dt>
<dd><p>character, the strand + or -</p></dd>


<dt><code>imputed</code></dt>
<dd><p>logical, whether imputed</p></dd>


<dt><code>info</code></dt>
<dd><p>numeric, the info score</p></dd>


<dt><code>q</code></dt>
<dd><p>numeric, the Q statistic for meta analysis results</p></dd>


<dt><code>q_p</code></dt>
<dd><p>numeric, the Q statistic P-value</p></dd>


<dt><code>i2</code></dt>
<dd><p>numeric, the I2 statistic</p></dd>


<dt><code>trait</code></dt>
<dd><p>character, the GWAS trait</p></dd>


<dt><code>id</code></dt>
<dd><p>character, the GWAS identifier</p></dd>


<dt><code>source</code></dt>
<dd><p>character, data source; either the file path, or "data.table" if loaded directly</p></dd>


<dt><code>correlation</code></dt>
<dd><p>matrix, a correlation matrix of signed R values between variants</p></dd>


<dt><code>map</code></dt>
<dd><p><code>ColumnMap</code>, a mapping of class ColumnMap</p></dd>


<dt><code>qc</code></dt>
<dd><p>list, a named list of filters; name is the filter expression and value is an integer vector of rows that fail the filter</p></dd>


</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

