<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>GWAS object — GWAS • genepi.utils</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS object — GWAS"><meta name="description" content="A GWAS object is a container for vectors of GWAS data, a correlation matrix, and
meta-data regarding quality control procedures applied at the point of object creation / data import."><meta property="og:description" content="A GWAS object is a container for vectors of GWAS data, a correlation matrix, and
meta-data regarding quality control procedures applied at the point of object creation / data import."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genepi.utils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a></li>
    <li><a class="dropdown-item" href="../articles/clumping.html">Clumping</a></li>
    <li><a class="dropdown-item" href="../articles/collider_bias.html">Collider bias</a></li>
    <li><a class="dropdown-item" href="../articles/ld_matrix.html">LD matrix</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Plotting</a></li>
    <li><a class="dropdown-item" href="../articles/rationale.html">Rationale and key points</a></li>
    <li><a class="dropdown-item" href="../articles/standardise_gwas.html">Standardise GWAS</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>GWAS object</h1>

      <div class="d-none name"><code>GWAS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A GWAS object is a container for vectors of GWAS data, a correlation matrix, and
meta-data regarding quality control procedures applied at the point of object creation / data import.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GWAS</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  map <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  drop <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fill <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fill_rsid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  missing_rsid <span class="op">=</span> <span class="st">"fill_CHR:BP"</span>,</span>
<span>  parallel_cores <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  dbsnp_dir <span class="op">=</span> <span class="fu">genepi.utils</span><span class="fu">::</span><span class="fu"><a href="which_dbsnp_directory.html">which_dbsnp_directory</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>beta_invalid <span class="op">=</span> <span class="st">"!is.infinite(beta) &amp; abs(beta) &lt; 20"</span>, eaf_invalid <span class="op">=</span></span>
<span>    <span class="st">"eaf &gt; 0 &amp; eaf &lt; 1"</span>, p_invalid <span class="op">=</span> <span class="st">"!is.infinite(p)"</span>, se_invalid <span class="op">=</span> <span class="st">"!is.infinite(se)"</span>,</span>
<span>    alleles_invalid <span class="op">=</span> <span class="st">"!is.na(ea) &amp; !is.na(oa)"</span>, chr_missing <span class="op">=</span> <span class="st">"!is.na(chr)"</span>, bp_missing</span>
<span>    <span class="op">=</span> <span class="st">"!is.na(bp)"</span>, beta_missing <span class="op">=</span> <span class="st">"!is.na(beta)"</span>, se_missing <span class="op">=</span> <span class="st">"!is.na(se)"</span>, p_missing <span class="op">=</span></span>
<span>    <span class="st">"!is.na(p)"</span>, eaf_missing <span class="op">=</span> <span class="st">"!is.na(eaf)"</span><span class="op">)</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_map <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dat">dat<a class="anchor" aria-label="anchor" href="#arg-dat"></a></dt>
<dd><p>a valid string file path to be read by <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> or a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table::data.table</a></code> object; the GWAS data source</p></dd>


<dt id="arg-map">map<a class="anchor" aria-label="anchor" href="#arg-map"></a></dt>
<dd><p>a valid input to the <code>ColumnMap</code> class constructor (a predefined map string id, a named list or character vector, or a ColumMap object)</p></dd>


<dt id="arg-drop">drop<a class="anchor" aria-label="anchor" href="#arg-drop"></a></dt>
<dd><p>a logical, whether to drop data source columns not in the column <code>map</code></p></dd>


<dt id="arg-fill">fill<a class="anchor" aria-label="anchor" href="#arg-fill"></a></dt>
<dd><p>a logical, whether to add (NAs) missing columns present in the column <code>map</code> but not present in the data source</p></dd>


<dt id="arg-fill-rsid">fill_rsid<a class="anchor" aria-label="anchor" href="#arg-fill-rsid"></a></dt>
<dd><p>either FALSE or a valid argument for the chrpos_to_rsid <code>build</code> argument, e.g. "b37_dbsnp156"</p></dd>


<dt id="arg-missing-rsid">missing_rsid<a class="anchor" aria-label="anchor" href="#arg-missing-rsid"></a></dt>
<dd><p>a string, how to handle missing rsids: one of "fill_CHR:BP", "fill_CHR:BP_OA_EA", "overwrite_CHR:BP",
"overwrite_CHR:BP:OA:EA", "none", or "leave"</p></dd>


<dt id="arg-parallel-cores">parallel_cores<a class="anchor" aria-label="anchor" href="#arg-parallel-cores"></a></dt>
<dd><p>an integer, number of cores to used for RSID mapping, default is maximum machine cores</p></dd>


<dt id="arg-dbsnp-dir">dbsnp_dir<a class="anchor" aria-label="anchor" href="#arg-dbsnp-dir"></a></dt>
<dd><p>path to the dbsnp directory of fst files see chrpos_to_rsid <code>dbsnp_dir</code> argument</p></dd>


<dt id="arg-filters">filters<a class="anchor" aria-label="anchor" href="#arg-filters"></a></dt>
<dd><p>a list of named strings, each to be evaluated as an expression to filter the data during the quality control steps (above)</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>a valid string file path to be read by <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> or a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table::data.table</a></code> object; the reference data</p></dd>


<dt id="arg-ref-map">ref_map<a class="anchor" aria-label="anchor" href="#arg-ref-map"></a></dt>
<dd><p>a valid input to the <code>ColumnMap</code> class constructor (a predefined map id (a string), a named list or character vector, or a
ColumMap object) defining at least columns <code>rsid</code> (or <code>chr</code>, <code>bp</code>), <code>ea</code>, <code>oa</code> and <code>eaf</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a logical, whether to print details</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>variable capture to be passed to the constructor, e.g. individual vectors for the slots, rather that <code>dat</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an S7 class genepi.utils::GWAS object</p>
    </div>
    <div class="section level2">
    <h2 id="slots">Slots<a class="anchor" aria-label="anchor" href="#slots"></a></h2>


<dl><dt><code>rsid</code></dt>
<dd><p>character, variant ID - usually in rs12345 format, however this can be changed with the <code>missing_rsid</code> argument</p></dd>


<dt><code>chr</code></dt>
<dd><p>character, chromosome identifier</p></dd>


<dt><code>bp</code></dt>
<dd><p>integer, base position</p></dd>


<dt><code>ea</code></dt>
<dd><p>character, effect allele</p></dd>


<dt><code>oa</code></dt>
<dd><p>character, other allele</p></dd>


<dt><code>eaf</code></dt>
<dd><p>numeric, effect allele frequency</p></dd>


<dt><code>beta</code></dt>
<dd><p>numeric, effect size</p></dd>


<dt><code>se</code></dt>
<dd><p>numeric, effect size standard error</p></dd>


<dt><code>p</code></dt>
<dd><p>numeric, p-value</p></dd>


<dt><code>n</code></dt>
<dd><p>integer, total number of samples</p></dd>


<dt><code>ncase</code></dt>
<dd><p>integer, number of cases</p></dd>


<dt><code>strand</code></dt>
<dd><p>character, the strand + or -</p></dd>


<dt><code>imputed</code></dt>
<dd><p>logical, whether imputed</p></dd>


<dt><code>info</code></dt>
<dd><p>numeric, the info score</p></dd>


<dt><code>q</code></dt>
<dd><p>numeric, the Q statistic for meta analysis results</p></dd>


<dt><code>q_p</code></dt>
<dd><p>numeric, the Q statistic P-value</p></dd>


<dt><code>i2</code></dt>
<dd><p>numeric, the I2 statistic</p></dd>


<dt><code>proxy_rsid</code></dt>
<dd><p>character, proxy variant ID</p></dd>


<dt><code>proxy_chr</code></dt>
<dd><p>character, proxy chromosome identifier</p></dd>


<dt><code>proxy_bp</code></dt>
<dd><p>integer, proxy base position</p></dd>


<dt><code>proxy_ea</code></dt>
<dd><p>character, proxy effect allele</p></dd>


<dt><code>proxy_oa</code></dt>
<dd><p>character, proxy other allele</p></dd>


<dt><code>proxy_eaf</code></dt>
<dd><p>numeric, proxy effect allele frequency</p></dd>


<dt><code>proxy_r2</code></dt>
<dd><p>numeric, proxy r2 with rsid</p></dd>


<dt><code>trait</code></dt>
<dd><p>character, the GWAS trait</p></dd>


<dt><code>id</code></dt>
<dd><p>character, the GWAS identifier</p></dd>


<dt><code>source</code></dt>
<dd><p>character, data source; either the file path, or "data.table" if loaded directly</p></dd>


<dt><code>correlation</code></dt>
<dd><p>matrix, a correlation matrix of signed R values between variants</p></dd>


<dt><code>map</code></dt>
<dd><p><code>ColumnMap</code>, a mapping of class ColumnMap</p></dd>


<dt><code>qc</code></dt>
<dd><p>list, a named list of filters; name is the filter expression and value is an integer vector of rows that fail the filter</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

