<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Liftover GWAS positions — lift • genepi.utils</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Liftover GWAS positions — lift"><meta name="description" content="Determine GWAS build and liftover to required build. This is the same function from the
GwasDataImport package, the only difference being that you can specify the build rather
than it trying to guess the build (which fails if you are trying to liftover
small segments of the genome)."><meta property="og:description" content="Determine GWAS build and liftover to required build. This is the same function from the
GwasDataImport package, the only difference being that you can specify the build rather
than it trying to guess the build (which fails if you are trying to liftover
small segments of the genome)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genepi.utils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.31</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/chrpos_to_rsid.html">CHR:POS to RSID mapping</a></li>
    <li><a class="dropdown-item" href="../articles/clumping.html">Clumping</a></li>
    <li><a class="dropdown-item" href="../articles/collider_bias.html">Collider bias</a></li>
    <li><a class="dropdown-item" href="../articles/ld_matrix.html">LD matrix</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Plotting</a></li>
    <li><a class="dropdown-item" href="../articles/rationale.html">Rationale and key points</a></li>
    <li><a class="dropdown-item" href="../articles/standardise_gwas.html">Standardise GWAS</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Liftover GWAS positions</h1>

      <div class="d-none name"><code>lift.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Determine GWAS build and liftover to required build. This is the same function from the
GwasDataImport package, the only difference being that you can specify the build rather
than it trying to guess the build (which fails if you are trying to liftover
small segments of the genome).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">lift</span><span class="op">(</span></span>
<span>  <span class="va">gwas</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Hg19"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Hg38"</span>,</span>
<span>  snp_col <span class="op">=</span> <span class="st">"snp"</span>,</span>
<span>  chr_col <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>  pos_col <span class="op">=</span> <span class="st">"pos"</span>,</span>
<span>  ea_col <span class="op">=</span> <span class="st">"ea"</span>,</span>
<span>  oa_col <span class="op">=</span> <span class="st">"oa"</span>,</span>
<span>  remove_duplicates <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gwas">gwas<a class="anchor" aria-label="anchor" href="#arg-gwas"></a></dt>
<dd><p>a data.table, or file path, chr, pos, snp name, effect allele, non-effect allele columns</p></dd>


<dt id="arg-from">from<a class="anchor" aria-label="anchor" href="#arg-from"></a></dt>
<dd><p>which build to lift from, one of c("Hg18", "Hg19", "Hg38")</p></dd>


<dt id="arg-to">to<a class="anchor" aria-label="anchor" href="#arg-to"></a></dt>
<dd><p>which build to lift over to, one of c("Hg18", "Hg19", "Hg38")</p></dd>


<dt id="arg-snp-col">snp_col<a class="anchor" aria-label="anchor" href="#arg-snp-col"></a></dt>
<dd><p>Name of SNP column name. Optional. Uses less certain method of matching if not available</p></dd>


<dt id="arg-chr-col">chr_col<a class="anchor" aria-label="anchor" href="#arg-chr-col"></a></dt>
<dd><p>Name of chromosome column name. Required</p></dd>


<dt id="arg-pos-col">pos_col<a class="anchor" aria-label="anchor" href="#arg-pos-col"></a></dt>
<dd><p>Name of position column name. Required</p></dd>


<dt id="arg-ea-col">ea_col<a class="anchor" aria-label="anchor" href="#arg-ea-col"></a></dt>
<dd><p>Name of effect allele column name. Optional. Might lead to duplicated rows if not presented</p></dd>


<dt id="arg-oa-col">oa_col<a class="anchor" aria-label="anchor" href="#arg-oa-col"></a></dt>
<dd><p>Name of other allele column name. Optional. Might lead to duplicated rows if not presented</p></dd>


<dt id="arg-remove-duplicates">remove_duplicates<a class="anchor" aria-label="anchor" href="#arg-remove-duplicates"></a></dt>
<dd><p>a logical, whether to remove duplicate IDs</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.table with updated position columns</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>https://github.com/MRCIEU/GwasDataImport</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicholas Sunderland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

